ContextQuery

danieldresser-ie opened the pull request:
Based heavily on Eric's ShaderQuery, though a lot simpler.

I started by adding an optional argument to PlugAlgo::canSetValueFromData for specifying a particular type to check ( let me know if you want this as an overload instead ).  This allows the value of exists to match when the default value is used, and should probably be used in other query nodes.

I did make a couple of arbitrary changes to to Eric's naming in my version - I renamed outPlugFromQuery/queryPlug to outPlugFromQueryPlug/queryPlugFromOutPlug, since that made the parallelism clearer to me.

The one worth talking about is probably that I omitted existsPlugFromQuery/valuePlugFromQuery.  ShaderQuery uses these sometimes to provide better error handling in the case of malformed outputs that are missing plugs - but doesn't seem to use these consistently - there are other places where it calls `getChild( g_existsPlugIndex )` directly, still resulting in undefined behaviour if the plug is missing.  In general in Gaffer, we don't really worrying about checking for plugs that we know exist - it is pretty easy to get undefined behaviour if you do something silly like deleting the "bound" child off a ScenePlug, and monkeying with the output of ShaderQuery isn't really much different.

I do wonder for exceptional cases though, if it wouldn't be better to have slightly better handling instead of just undefined behaviour - is there any reason why GraphComponent.getChild( index ) shouldn't throw if the index is out of bounds?

Other than that aside about corner cases, hopefully this is fairly straightforward.  The main thing to watch for in review is that I haven't copied something I shouldn't - the whole UI is pretty well copied from ShaderQuery, with minimal changes.  Seems to be working fine though.

johnhaddon commented on src/Gaffer/ContextQuery.cpp:
```diff
@@ -0,0 +1,342 @@
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new ArrayPlug( "queries" ) );
+	addChild( new ArrayPlug( "out", Plug::Direction::Out ) );
```
I think we need to disable the `resizeWhenInputsChange` behaviour of ArrayPlug here (via the constructor argument), so that `addQuery()` is in sole control of queries. I _think_ this will fix the following bug :

1. Make ContextQuery.
2. Add a string query via the UI.
3. Right click on the `name` widget for the query, and choose `Create Spreadsheet` or `Create Python Expression`.
4. Note that a new query has appeared, along with this message in the shell : `ERROR : gui.scripts.ScriptNode.ContextQuery::affects() : ContextQuery : "outPlug" is missing`.

I believe the same applies to ShaderQuery.

johnhaddon replied:
This one is worth doing for ShaderQuery on `0.61_maintenance` @ericmehl.

johnhaddon commented on python/GafferUI/ContextQueryUI.py:
```diff
@@ -0,0 +1,408 @@
+
+		result = IECore.MenuDefinition()
+
+		for item in [
+			Gaffer.BoolPlug,
+			Gaffer.FloatPlug,
+			Gaffer.IntPlug,
+			"NumericDivider",
+			Gaffer.StringPlug,
+			"StringDivider",
+			Gaffer.V2iPlug,
+			Gaffer.V3iPlug,
+			Gaffer.V2fPlug,
+			Gaffer.V3fPlug,
+			"VectorDivider",
+			Gaffer.Color3fPlug,
+			Gaffer.Color4fPlug,
```
Let's extend this list to include all the types that the ContextVariablesUI allows - arrays etc. At some point we need to unify all this "menu for making plugs" code, but don't worry about that now. We can just manually match the list from ContextVariablesUI.

johnhaddon commented on python/GafferUI/ContextQueryUI.py:
```diff
@@ -0,0 +1,408 @@
+##########################################################################
+
+
+def __plugPopupMenu( menuDefinition, plugValueWidget ) :
+
+	readOnlyUI = plugValueWidget.getReadOnly()
+	plug = plugValueWidget.getPlug().ancestor( Gaffer.NameValuePlug )
+
+	if plug is not None :
+
+		if len( menuDefinition.items() ) :
+			menuDefinition.append( "/DeleteDivider", { "divider" : True } )
+
+		menuDefinition.append( "/Delete", { "command" : functools.partial( __deletePlug, plug ), "active" : not readOnlyUI and not Gaffer.MetadataAlgo.readOnly( plug ) } )
```
This has copied the AttributeQuery bug that was fixed here : https://github.com/GafferHQ/gaffer/pull/4678. We'll need the same fix for this...

johnhaddon commented on python/GafferUI/ContextQueryUI.py:
```diff
@@ -0,0 +1,408 @@
+			The default value to output if the variable does not exist.
+			""",
+
+			"spreadsheet:columnName", functools.partial( __getLabel, parentPlug = "queries", sanitise = True ),
```
I don't know what we're hoping to achieve here - I guess we're trying to get the spreadsheet column to be named after `query["name"].getValue()`, in a similar way to what we do for ArnoldAttributes etc? In any case, it's creating this error when trying to create a spreadsheet (via the "Value and Switch") option :

```
  File "/home/john/dev/build/gaffer62Python3/python/GafferUI/SpreadsheetUI/_Algo.py", line 56, in createSpreadsheet
    addColumn( spreadsheet, plug )
  File "/home/john/dev/build/gaffer62Python3/python/GafferUI/SpreadsheetUI/_Algo.py", line 88, in addColumn
    columnName = Gaffer.Metadata.value( plug, "spreadsheet:columnName" ) or plug.getName()
  File "/home/john/dev/build/gaffer62Python3/python/GafferUI/ContextQueryUI.py", line 56, in __getLabel
    qPlug = n.queryPlugFromOutPlug( plug )
IECore.Exception: ContextQuery : Plug not in hierarchy.
```

I don't think we're likely to drive this with a spreadsheet, so rather than worry about it, I think we'd be better off just removing the metadata for now (same for ShaderQuery, since I assume that's where it came from).

ericmehl replied:
> I guess we're trying to get the spreadsheet column to be named after query["name"].getValue()

Yes, I had included it for `ShaderQuery` as part of the effort to make the UI a little more friendly - the partial function is called for tooltips as well.

Strangely it's working for adding the (not terribly useful) `name` plug to a `Spreadsheet`, but not `value`. If it's worth moving the column label handling to `queries.*.value` metadata, I'm happy sorting out the bug as part of `ShaderQuery`.

johnhaddon replied:
I'm still not sure why we'd be adding these to a spreadsheet, or what would be useful in that case :

- If we're adding the whole NameValuePlug (the case that errors now), then it's not meaningful to name the column after `query["name"].getValue()`, because the name will be dynamic (driven by the spreadsheet).
- If we're adding just the `query["name"]` to the spreadsheet, the same applies.
- I can't think of a plausible reason I'd add `query["value"]` to a spreadsheet - I don't know why I'd want a dynamic default value.

Unless we have a good use case, I'm inclined to save ourselves the bother and just remove the spreadsheet metadata.

johnhaddon commented on python/GafferUI/ContextQueryUI.py:
```diff
@@ -0,0 +1,408 @@
+	"""
+	Queries variables from the current context, creating outputs for each variable.
+	""",
+
```
I think we probably want to style the ContextQuery like expression, animation, random and spreadsheet - see https://github.com/GafferHQ/gaffer/blob/main/python/GafferUI/ExpressionUI.py#L56-L61 for relevant metadata.

johnhaddon replied:
Turning off `resizeWhenInputsChange` for the ArrayPlug has caused the appearance of an unwanted `+` button when using the StandardNodeGadget :

![image](https://user-images.githubusercontent.com/1133871/172198109-27b3a10a-45a8-4c2f-80f5-77236d4ac171.png)

Ideally I guess we'd have a custom `+` button that called `addQuery()` appropriately, but I think the NodeEditor UI for that is sufficient for now. We can disable the unwanted button by registering this metadata against the `out` plug :

```"noduleLayout:customGadget:addButton:gadgetType", ""```

johnhaddon commented on python/GafferUI/ContextQueryUI.py:
```diff
@@ -0,0 +1,408 @@
+##########################################################################
+
+
+def __plugPopupMenu( menuDefinition, plugValueWidget ) :
```
We discussed having a `Create Context Query...` menu item on plug popup menus, in the same vein as `Create Python Expression...`. I think it's worth doing that now, as it'll make it a lot easier for the less technical user to wire one up.

johnhaddon reviewed (changes requested):
Thanks Daniel - this is going to be really handy, and should lower the bar for less technical users to make the most of varying contexts.

Lots of comments inline, but I think they should all be pretty straightforward.

> The one worth talking about is probably that I omitted existsPlugFromQuery/valuePlugFromQuery.

I'm not sure this is purely about error handling for missing plugs. Those functions also make it easier to wire up the node from C++. Which reminds me, we should replace ImageView's internal `V2fContextVariable` node with this one. When doing that, maybe have a think about whether `valuePlugFromQuery()` is justified from that perspective?

> is there any reason why GraphComponent.getChild( index ) shouldn't throw if the index is out of bounds?

Probably not, but let's call that out of scope for this PR...

Cheers...
John

ericmehl commented on src/Gaffer/ContextQuery.cpp:
```diff
@@ -0,0 +1,342 @@
+	NameValuePlugPtr childQueryPlug = new NameValuePlug(
+		"",
+		plug->createCounterpart( "query", Gaffer::Plug::Direction::In ),
+		"query1",
```
If so, should it be applied to `ShaderQuery` as well? My original reasoning there was naming the user-facing plugs a little more naturally, but `query0` is probably familiar to most people writing expressions and such...

johnhaddon replied:
My eventual aim is to eliminate names entirely for the children of array plugs, so we only use `[index]` to access them. In the meantime, I think there's value in making sure that `array[N]` and `array["queryN"]` agree on what N is.

Technically I guess applying this to ShaderQuery would be a breaking change (although it shouldn't affect serialisation, because we _do_ use indices there). So if you could make that change for `main` @ericmehl, that'd be great.

ericmehl commented on src/Gaffer/ContextQuery.cpp:
```diff
@@ -0,0 +1,342 @@
+	);
+	childQueryPlug->namePlug()->setValue( parameter );
+
+	ValuePlugPtr newOutPlug = new ValuePlug( "out1", Gaffer::Plug::Direction::Out );
```
Also as above regarding renaming `out1` in `ShaderQuery`.

danieldresser-ie commented:
OK, I think I've addressed all feedback, including adding back existsPlugFromQueryPlug/valuePlugFromQueryPlug to make it easier to use ContextQuery in ImageView.

Couple of things worth talking about:

I've made the new accessors non-const ... this feels a little weird, and doesn't match ShaderQuery, but they are there so that other nodes can setInput from them, and setInput requires a non-const plug pointer.

You had previously said that handling out-of-bounds indices in GraphComponent.getChild is out of scope, but when I went to copy ShaderQuery's existsPlugFromQuery and its test, I noticed that as far as I can tell, this test is relying on undefined behaviour:  https://github.com/GafferHQ/gaffer/blob/main/python/GafferSceneTest/ShaderQueryTest.py#L113
As far as I can tell, currently, this accesses an out of range value in GraphComponent::m_children and interprets it as a pointer.  I guess that when it tries to access the typeId of this unallocated memory, it doesn't match the typeId for ValuePlug, and therefore returns null.  Since we already have code using this behaviour, and it feels extremely reasonable ( getChild already returns null if the type doesn't match, so why not return null for out of bound as well ), it seems like the quickest fix is to actually implement getChild to return null for out of bound.  I don't really like leaving the code relying on undefined behaviour ... I guess if we don't want to touch GraphComponent, I could just remove the tests from ShaderQuestTest and ContextQueryTest

ericmehl commented:
Agreed on @danieldresser interpretation of what `ShaderQuery` is currently testing for, including relying on undefined behavior of `getChild` in the tests. 

I don't feel like I have enough of a handle on the ramifications of changing `getChild` to return `null` for out of bound, but another easy solution would be to have `ShaderQuery` check for bounds in `existsPlugFromQuery` and `valuePlugFromQuery` like it does in `outPlugFromQuery` and throw an exception instead of return a null value.

The test is covering what is hopefully a pretty obscure case of the `existsPlug` or `valuePlug` being forcibly removed from the parent, which does seem like it would be deserving of an exception.

johnhaddon commented:
I've noted a couple of very minor last minute UI things in the inline conversations. Would be good to get those done before we merge.

> I've made the new accessors non-const ... this feels a little weird, and doesn't match ShaderQuery, but they are there so that other nodes can setInput from them, and setInput requires a non-const plug pointer.

Our typical convention for this is to provide both const and non-const versions I think, returning const and non-const pointers respectively. But if we don't need the const version right now, and doesn't seem that important.

> You had previously said that handling out-of-bounds indices in GraphComponent.getChild is out of scope, but when I went to copy ShaderQuery's existsPlugFromQuery and its test, I noticed that as far as I can tell, this test is relying on undefined behaviour

Oh, you're right! My mistake - I hadn't spotted that during the ShaderQuery review. We should just remove that test - it's not our intention to protect against that kind of behaviour.

> it seems like the quickest fix is to actually implement getChild to return null for out of bound.

I can't think of any precedent for an out-of-array-bounds accessor returning null - it seems to me it should either be undefined behaviour or an exception. I still think this is out of scope - let's just remove the test.
